common/serde-reflection/src/trace.rs

```diff
@@ -2,473 +2,233 @@
 
-    fn serialize_field<T>(&mut self, value: &T) -> Result<()>
+    /// Same as `trace` for seeded deserialization.
+    pub fn trace_type_with_seed<'de, T>(&'de mut self, seed: T) -> Result<Format>
     where
-        T: ?Sized + Serialize,
+        T: DeserializeSeed<'de> + Clone,
     {
-        let format = value.serialize(&mut *self.tracer)?;
-        self.fields.push(format);
-        Ok(())
-    }
-
-    fn end(self) -> Result<Format> {
-        let format = Format::TupleStruct(self.fields);
-        self.tracer.record(self.name, format)
+        loop {
+            let format = self.trace_type_once_with_seed(seed.clone())?;
+            if let Format::TypeName(name) = &format {
+                if self.incomplete_enums.contains(name) {
+                    self.incomplete_enums.remove(name);
+                    continue;
+                }
+            }
+            return Ok(format);
+        }
     }
-}
-
-pub struct TupleVariantTracer<'a> {
-    tracer: &'a mut Tracer,
-    name: &'static str,
-    variant_index: u32,
-    variant_name: &'static str,
-    fields: Vec<Format>,
-}
-
-impl<'a> ser::SerializeTupleVariant for TupleVariantTracer<'a> {
-    type Ok = Format;
-    type Error = Error;
 
-    fn serialize_field<T>(&mut self, value: &T) -> Result<()>
+    /// Compute a default value for `T`.
+    /// Return true if `T` is an enum and more values are needed to cover
+    /// all variant cases. Otherwise similar to `trace_type_once`.
+    pub fn sample_type_once<'de, T>(&mut self) -> Result<(T, bool)>
     where
-        T: ?Sized + Serialize,
+        T: Deserialize<'de>,
     {
-        let format = value.serialize(&mut *self.tracer)?;
-        self.fields.push(format);
-        Ok(())
-    }
-
-    fn end(self) -> Result<Format> {
-        let variant = VariantFormat::Tuple(self.fields);
-        self.tracer
-            .record_variant(self.name, self.variant_index, self.variant_name, variant)
-    }
-}
-
-pub struct MapTracer<'a> {
-    tracer: &'a mut Tracer,
-    key: Format,
-    value: Format,
-}
-
-impl<'a> ser::SerializeMap for MapTracer<'a> {
-    type Ok = Format;
-    type Error = Error;
-
-    fn serialize_key<T>(&mut self, key: &T) -> Result<()>
+        let mut format = Format::Unknown;
+        let deserializer = Deserializer::new(&mut *self, &mut format);
+        let value = T::deserialize(deserializer)?;
+        let has_more = match &format {
+            Format::TypeName(name) => {
+                if self.incomplete_enums.contains(name) {
+                    self.incomplete_enums.remove(name);
+                    true
+                } else {
+                    false
+                }
+            }
+            _ => false,
+        };
+        Ok((value, has_more))
+    }
+
+    /// Same as sample_type_once but if `T` is an enum, we repeat the process
+    /// until all variants are covered. Otherwise similar to `trace_type`.
+    pub fn sample_type<'de, T>(&mut self) -> Result<Vec<T>>
```

Comment:
Ah! Ok so i was misreading the above function so my comment above probably is meaningless.

The bool is used to tell you if you need to run through more variants, not that they can't be produced. right?

Reply:
right